src/url-replacements.js

```diff
@@ -239,6 +242,14 @@ class UrlReplacements {
       return this.getTimingData_('navigationStart',
           'domContentLoadedEventStart');
     });
+
+    // Returns an identifier for the viewer.
+    this.set_('VIEWER', () => {
+      const viewerUrl = viewerFor(this.win_).getParam(VIEWPORT_ORIGIN_PARAM_) ||
```

Comment:
`messagingOrigin_` is easy - you simply need to wrap it in a promise. E.g. you will have an open promise that will be resolved in the same place where `messagingOrigin_` is set.

RE: viewer vs referrer - please touch base with @jridgewell - he dealt with this problem. I believe what he did was to translate a set of referrers to viewers. But I could be wrong.